web/tests/user_groups.test.cjs

```diff
@@ -755,3 +755,396 @@ run_test("is_subgroup_of_target_group", () => {
 
     assert.ok(!user_groups.is_subgroup_of_target_group(students.id, all.id));
 });
+
+run_test("group_has_permission", () => {
+    const admins = {
+        name: "Administrators",
+        id: 1,
+        members: new Set([1]),
+        is_system_group: false,
+        direct_subgroup_ids: new Set([]),
+    };
+    const moderators = {
+        name: "Moderators",
+        id: 2,
+        members: new Set([2]),
+        is_system_group: false,
+        direct_subgroup_ids: new Set([1]),
+    };
+    const all = {
+        name: "Everyone",
+        id: 3,
+        members: new Set([3, 4]),
+        is_system_group: false,
+        direct_subgroup_ids: new Set([2, 4]),
+    };
+    const students = {
+        name: "Students",
+        id: 4,
+        members: new Set([5]),
+        is_system_group: false,
+        direct_subgroup_ids: new Set([]),
+    };
```

Comment:
This test could benefit from some comments laying out the scenarios. I also think we will want to build something like `web/tests/lib/example_user.cjs` for making a collection of sample groups, including the system groups (with maybe some hackery to make the membership be computed from what's in `people.ts` based on role), so you can just call a couple functions and get some useful collections of groups to work with in tests. But that's definitely out of scope for this PR, and is mostly important for other test suites.

None of that is urgent though, just a good thing to do when you want a break.